FP_analysis_of_reviews_from_Amazon/Task_9_1.py

```diff
@@ -20,25 +21,26 @@ def bad_comment(data):
     the application which received the most useless message;
     """
     return min([review for review in data if review["helpful"][1] != 0],
-               key=lambda review: review["helpful"][0] / review["helpful"][1])
+               key=lambda review: review["helpful"][0] / review["helpful"][1],
+               default={"asin": None, "helpful": [float('inf'), 1], "reviewText": None})
 
 
-def nonanalys_data(reviews):
+def nonanalys_data(data):
     """function returns the number of records that cannot be processed for every point above.
 
     1.5. Task_9_1: to create file general-stats.cvs containing information about
     the number of records that cannot be processed for every point above.
     """
     def count_unanalyzed(*required_keys):
         return len(
-            [review for review in reviews
-             if all(key not in review.keys() for key in required_keys)]
+            [review for review in data
+             if any(key not in review.keys() for key in required_keys)]
```

Comment:
let's use [generator expressions](https://www.python.org/dev/peps/pep-0289/) inside all `len`, `min`, `sum` etc.

Reply:
I read documentation and understood idea, it seems very effectively. But I don't know how to use this in my code. For example, I have 'best_comment' function:
```
def best_comment(data):
    return max(data, key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
I expect that I will receive something like that:
{
      asin: ...,
      reviewerID:....
      helpful: [found max value, ....],
      textreview: ...,
      ...
}
or default value, if max value isn't exist.
But, if I use generator expression in this function:
```
def best_comment(data):
    return max(review["helpful"][0] for review in data)
```
I will receive integer, which equal max value. Such variant isn't suit for me.
How I understood, this variant isn't generator expression and equal the first variant:
```
def best_comment(data):
    return max((review for review in data), key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
Just in case, I redid everything like third variant, except len, because I received exception 'generation object doesn't have attribute len'.

Reply:
Lemmy jump in. @Tasenga in your last example:
```
def best_comment(data):
    return max((review for review in data), key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
`(review for review in data)` it's actually a generator expression, but you applied it in the not best way. The `data` will actually be iterated twice: by generator and by max func, which is overhead in this particular case.  

Let's better consider this example:
```
def count_unanalyzed(*required_keys):
        return len(
            [review for review in data
             if any(key not in review.keys() for key in required_keys)]
        )
```

How generators can be applied here?

Reply:
I returned previous implementation of best_comment.
I replaced len with sum everywhere if it was possible, such an example:
```
    def count_unanalyzed(*required_keys):
        return sum(
            1 for review in data
             if any(key not in review.keys() for key in required_keys)
        )
```

Reply:
I'm sorry I was not very precise. This code:
```
return max(data, key=lambda review: review["helpful"][0],
                      default={"asin": None, "helpful": [None, None], "reviewText": None})
```
is absolutely fine. The idea was to avoid list comprehensions and use generator expressions instead of them where possible. Not literally *all* len, min and sum needed to be editied, sorry for that phrase

Reply:
`sum(1 for ...)` is a good idea, this'll save memory

Reply:
@Tasenga I cannot see `sum(1 ...` being used in your code, please update

Reply:
I'm sorry, updated. It is because branch has been renamed